src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -0,0 +1,1050 @@
+            _cpuPtr->microcodeRom.fetchMicroop(next_pc.microPC(), curMacroOp);
+
+        DPRINTF(SDCPUInstEvent,
+                "Decoded ROM microop (seq %d) - %#x : %s\n",
+                seq_num,
+                next_pc.microPC(),
+                static_inst->disassemble(next_pc.microPC()).c_str());
+
+        shared_ptr<InflightInst> dynamic_inst_ptr =
+            make_shared<InflightInst>(getThreadContext(), isa, &memIface,
+                                      seq_num, next_pc, static_inst);
+
+        inflightInsts.push_back(dynamic_inst_ptr);
+
+        issueInstruction(dynamic_inst_ptr);
+
+        return;
+    }
+
+    if (curMacroOp) {
+        const StaticInstPtr static_inst =
+            curMacroOp->fetchMicroop(next_pc.microPC());
+
+        DPRINTF(SDCPUInstEvent,
+                "Decoded microop (seq %d) - %#x : %s\n",
+                seq_num,
+                next_pc.microPC(),
+                static_inst->disassemble(next_pc.microPC()).c_str());
+
+        shared_ptr<InflightInst> dynamic_inst_ptr =
+            make_shared<InflightInst>(getThreadContext(), isa, &memIface,
+                                      seq_num, next_pc, static_inst);
+
+        inflightInsts.push_back(dynamic_inst_ptr);
+
+        issueInstruction(dynamic_inst_ptr);
+
+        return;
+    }
+
+    shared_ptr<InflightInst> dynamic_inst_ptr =
+        make_shared<InflightInst>(getThreadContext(), isa, &memIface, seq_num,
+                                  next_pc);
+
+    inflightInsts.push_back(dynamic_inst_ptr);
+
+    fetchOffset = 0;
+    attemptFetch(weak_ptr<InflightInst>(dynamic_inst_ptr));
+}
+
+void
+SDCPUThread::attemptFetch(weak_ptr<InflightInst> inst)
+{
+    const shared_ptr<InflightInst> inst_ptr = inst.lock();
+    if (!inst_ptr || inst_ptr->isSquashed()) {
+        // No need to do anything for an instruction that has been squashed.
+        return;
+    }
+
+    DPRINTF(SDCPUThreadEvent, "attemptFetch()\n");
+
+    // TODO check for interrupts
+    // NOTE: You cannot take an interrupt in the middle of a macro-op.
+    //      (How would you recover the architectural state since you are in a
+    //       non-architectural state during a macro-op?)
+    //      Thus, we must check instruction->isDelayedCommit() and make sure
+    //      that interrupts only happen after an instruction *without* the
+    //      isDelayedCommit flag is committed.
+    // TODO something something PCEventQueue?
+
+    TheISA::PCState pc_value = inst_ptr->pcState();
+
+    // Retrieve the value of the speculative PC as a raw Addr value, aligned
+    // to MachInst size, but offset by the amount of data we've fetched thus
+    // far, since multiple fetches may be needed to feed one decode.
+    const Addr pc_addr = pc_value.instAddr();
+    const Addr req_vaddr = (pc_addr & BaseCPU::PCMask) + fetchOffset;
+
+    // If HIT in fetch buffer, we can skip sending a new request, and decode
+    // from the buffer immediately.
+    if (fetchBufBase
```

Comment:
I think the fetch buffer should be in the CPU, not in the thread. This way you can check for bandwidth/latency on all fetches.